proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+```
+
+instead of specifying directly the underlying byte-array:
+
+```
+.Map[ RET_ADDR      := 0  ]
+    [ RET_ADDR + 1  := 0  ]
+    ...
+    [ RET_ADDR + 31 := 1 ]
+```
+
+### Storage
+
+The storage accommodates permanent data such as the `balances` map.
+
+A map is laid out in the storage where the map entries are scattered over the entire storage space for which the hash of each key is used to determine the location.
+
+The detailed mechanism of calculating the location varies by compilers. In Viper, for example, `map[key1][key2]` is stored at the location:
+
+```
+hash(hash(pos(map)) + key1) + key2
+```
+
+where `pos(map)` is the position of `map` appearing in the program, and `+` is the addition modulo `2^256`, while in Solidity, it is stored at:
+
+```
+hash(key2 ++ hash(key1 ++ pos(map)))
+```
+
+where `++` is the byte-array concatenation.
+
+The hashed location abstraction allows to uniformly specify the locations in a form parameterized by the underlying compilers. For example, the location of `map[key1][key2]` can be specified as follows, where `{lang}` is a place-holder to be replaced by the name of the compiler.
+
+```
+#hashedLocation({lang}, pos(map), key1 key2)
+```
+
+This abstraction makes the specification independent of the underlying compilers, enabling it to be reused for differently compiled programs.
+
+
+### Gas Consumption
+
+Regarding the gas consumption of the program, the maximum gas amount is specified to ensure that the program does not consume more gas than the limit.
+
+The verifier proves that the actual gas consumption is less than the provided limit, and also reports the exact amount of gas consumed during the execution. Indeed, it reports a set of the amounts since the gas consumption varies depending on the context (i.e., the input parameter values and the state of the storage). This provides the most precise gas analysis.
+
+
+
+We present the specification for each function defined in the ERC20 standard. Here we focus on explaining the EVM-specific detailed behaviors, referring to the ERC20-K specification for the high-level logic.
+
+
+totalSupply
+-----------
+
+```{.k .erc20-spec .hkg-spec}
+[totalSupply]
+```
+
+`<k>` cell specifies that the execution eventually reaches RETURN instruction meaning that the program will successfully terminate. The RETURN instruction says that a 32-byte return value will be stored in the memory at the location RET_ADDR.
+
+```{.k .erc20-spec .hkg-spec}
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("totalSupply", .TypedArgs)
+```
+
+`<localMem>` cell specifies that the local memory is empty in the beginning, but in the end, it will contain some contents including the return value `TOTAL`, the total supply of the tokens.
+
+The other entries are represented by the wildcard symbol `_`, meaning that they are not relevant to the functional correctness of the program.
+
+```{.k .erc20-spec .hkg-spec}
+localMem:
+      .Map
```

Comment:
Is the local memory empty for sure when a function is executed?  Is this because you use the execution and not the caller's context when specifying the property, so the caller's local memory has already been stacked?  Is that stacking code then ever verified?

Reply:
Yes, and will be done next version as I said in the above.

Reply:
nice, but again, a short sentence about it now?  Later on, when we do things in the caller's context, we will go back and modify the text.  We will want to revise the entire text anyway.